allennlp/commands/elmo.py

```diff
@@ -0,0 +1,203 @@
+            instances.append(instance)
+
+        dataset = Batch(instances)
+        vocab = Vocabulary()
+        dataset.index_instances(vocab)
+        return dataset.as_tensor_dict()['elmo']['character_ids']
+
+    def batch_to_embeddings(self, batch: List[List[str]]) -> Tuple[torch.Tensor, torch.Tensor]:
+        """
+        :param batch: a list of tokenized sentences
+        :return: a tuple of tensors, the first representing activations (batch_size, 3, num_timesteps, 1024) and
+        the second a mask (batch_size, num_timesteps)
+        """
+        character_ids = self.batch_to_ids(batch)
+        if self.cuda_device >= 0:
+            character_ids = character_ids.cuda(cuda_device=self.cuda_device)
+
+        bilm_output = self.elmo_bilm(character_ids)
+        layer_activations = bilm_output['activations']
+        mask_with_bos_eos = bilm_output['mask']
+
+        # without_bos_eos is a 3 element list of pairs of (batch_size, num_timesteps, dim) tensors
+        without_bos_eos = [remove_sentence_boundaries(layer, mask_with_bos_eos)
+                           for layer in layer_activations]
+        # Converts a list of pairs (activation, mask) tensors to a single tensor of activations.
+        activations = torch.cat([ele[0].unsqueeze(1) for ele in without_bos_eos], dim=1)
+        # The mask is the same for each ELMo vector, so just take the first.
+        mask = without_bos_eos[0][1]
+
+        return activations, mask
+
+    def embed_sentence(self, sentence: List[str]) -> torch.Tensor:
+        """
+        Computes the ELMo embeddings for a single tokenized sentence.
+        :param sentence: a tokenized sentence
+        :return: a tensor containing the ELMo vectors
+        """
+        return self.embed_batch([sentence])[0]
+
+    def embed_batch(self, batch: List[List[str]]) -> List[torch.Tensor]:
+        """
+        Computes the ELMo embeddings for a batch of tokenized sentences.
+        :param batch: a list of tokenized sentences
+        :return: a list of tensors, each representing the ELMo vectors for the input sentence at the same index
+        """
+        elmo_embeddings = []
+
+        embeddings, mask = self.batch_to_embeddings(batch)
+        for i in range(len(batch)):
+            length = int(mask[i, :].sum())
+            sentence_embeds = embeddings[i, :, :length, :].data.cpu().numpy()
+            elmo_embeddings.append(sentence_embeds)
+
+        return elmo_embeddings
+
+    def embed_sentences(self, sentences: Iterable[List[str]], batch_size: int) -> Iterable[torch.Tensor]:
+        """
+        Computes the ELMo embeddings for a iterable of sentences.
+        :param sentences: an iterable of tokenized sentences
+        :param batch_size: the number of sentences ELMo should process at once
+        :return: a list of tensors, each representing the ELMo vectors for the input sentence at the same index
+        """
+        for batch in lazy_groups_of(iter(sentences), batch_size):
```

Comment:
Can you add a tqdm wrapper to this? It would be nice to see the updates, because it could take a little while.

Reply:
I'd rather not.  I would prefer to leave it up to the caller to wrap the returned iterator.  Otherwise this method forces a tqdm that is not configurable.

As an example of a caller wrapping the iterator in a tqdm see `embed_file`.